### PR TITLE
[API]Fixed endpoints related to order-items that returns 500 error

### PR DESCRIPTION
**Pull request by arti0090**
| Q               | A
| --------------- | -----
| Branch?         |  1.8
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | not sure
| Deprecations?   | not sure
| License         | MIT

### CODE REVIEW

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    In order to see details about my order
+    As a visitor
+    I want to be able to see my cart summary
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "$12.54"
+        And there are 10 units of product "T-shirt banana" available in the inventory
+
+
```
Comment (AdamKasp) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:12]:
```suggestion
```

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    @api
+    Scenario: Viewing information about empty cart
+        When I add 5 of them to my cart
+        When I check items of my cart
```
Comment (AdamKasp) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:16]:
```suggestion
        And I check items in my cart
```

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    I want to be able to see my cart summary
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "$12.54"
+        And there are 10 units of product "T-shirt banana" available in the inventory
+
+
+    @api
+    Scenario: Viewing information about empty cart
```
Comment (AdamKasp) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:14]:
```suggestion
    Scenario: Viewing content of my cart
```
or sth similar, in this test you don't check emptiness of cart :)

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -323,6 +323,31 @@ private function getProductVariantForItem(array $item): Response
...
+    /**
+     * @Then /^my cart should have (\d+) items of (product "([^"]+)")$/
+     */
+    public function myCartShouldHaveItems(int $quantity, ProductInterface $product)
+    {
+        $response = $this->cartsClient->getLastResponse();
+
+        $items = json_decode($response->getContent(), true)['hydra:member'];
+
+        foreach($items as $item) {
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:344]:
```suggestion
        foreach ($items as $item) {
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 5]
```diff
@@ -323,6 +323,31 @@ private function getProductVariantForItem(array $item): Response
...
+            if($item['productName'] === $product->getName()){
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:345]:
```suggestion
            if ($item['productName'] === $product->getName()){
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderItem.xml [hunk 6]
```diff
@@ -16,7 +16,11 @@
            xsi:schemaLocation="https://api-platform.com/schema/metadata https://api-platform.com/schema/metadata/metadata-2.0.xsd"
 >
     <resource class="%sylius.model.order_item.class%" shortName="OrderItem">
-        <collectionOperations />
```
Comment (AdamKasp) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderItem.xml:19]: why did you delete it?

**Comment by AdamKasp** [comment]
And it should be open to 1.8

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 7]
```diff
@@ -0,0 +1,16 @@
+@shopping_cart
+Feature: Viewing a cart summary
+    In order to see details about my order
+    As a visitor
```
Comment (GSadee) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:4]:
```suggestion
    As a Visitor
```

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 8]
```diff
@@ -0,0 +1,16 @@
+@shopping_cart
```
Comment (GSadee) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:1]: The file name could be adjusted to the feature name

File: features/cart/shopping_cart/resolving_cart_order_items.feature [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+Feature: Viewing a cart summary
+    In order to see details about my order
+    As a visitor
+    I want to be able to see my cart summary
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "T-shirt banana" priced at "$12.54"
+        And there are 10 units of product "T-shirt banana" available in the inventory
```
Comment (GSadee) [inline features/cart/shopping_cart/resolving_cart_order_items.feature:10]: Is this step needed in this scenario?

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 10]
```diff
@@ -323,6 +323,31 @@ private function getProductVariantForItem(array $item): Response
         return $this->cartsClient->getLastResponse();
     }
 
+    /**
+     * @When /^I check items in my (cart)$/
+     */
+    public function iCheckItemsOfMyCart(string $tokenValue)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:329]:
```suggestion
    public function iCheckItemsOfMyCart(string $tokenValue): void
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 11]
```diff
@@ -323,6 +323,31 @@ private function getProductVariantForItem(array $item): Response
...
+    {
+        $request = Request::customItemAction(null,'orders', $tokenValue, HttpRequest::METHOD_GET, 'items');
+        $this->cartsClient->executeCustomRequest($request);
+    }
+
+    /**
+     * @Then /^my cart should have (\d+) items of (product "([^"]+)")$/
+     */
+    public function myCartShouldHaveItems(int $quantity, ProductInterface $product)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:338]:
```suggestion
    public function myCartShouldHaveItems(int $quantity, ProductInterface $product): void
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderItem.xml [hunk 12]
```diff
@@ -25,7 +31,12 @@
         </itemOperations>
 
         <property name="id" identifier="true" writable="false" />
-        <property name="adjustments" writable="false">
+        <property name="order" readable="true" />
+        <property name="productName" readable="true" />
+        <property name="product" readable="true">
+            <subresource resourceClass="%sylius.model.product.class%" />
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderItem.xml:37]: I'm not sure if the product should be a subresource here. Could you add a screenshot of an API response?

Comment (arti0090) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/OrderItem.xml:37]: Indeed there is no need, I will delete those lines with next changes.

File: features/cart/shopping_cart/resolving_cart_order_items.feature (hunk 8 above)
Comment (GSadee) [reply features/cart/shopping_cart/resolving_cart_order_items.feature:1]: To be honest, I thought about changing file name, not feature name, because in this scenario you just view items in the cart. `Resolving` doesn't look like an appropriate word here

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 13]
```diff
@@ -352,4 +372,14 @@ private function changeQuantityOfOrderItem(string $orderItemId, int $quantity, s
...
     }
+
+    private function hasItemWithNameAndQuantity(Response $response, string $productName, int $quantity): bool
+    {
+        $items = json_decode($response->getContent(), true)['hydra:member'];
+
+        foreach ($items as $item) {
+            return ($item['productName'] === $productName) && $item['quantity'] === $quantity;
+        }
+        return false;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:383]: Missing blank line before return

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 14]
```diff
@@ -352,4 +372,14 @@ private function changeQuantityOfOrderItem(string $orderItemId, int $quantity, s
 
         $this->cartsClient->executeCustomRequest($request);
     }
+
+    private function hasItemWithNameAndQuantity(Response $response, string $productName, int $quantity): bool
+    {
+        $items = json_decode($response->getContent(), true)['hydra:member'];
+
+        foreach ($items as $item) {
+            return ($item['productName'] === $productName) && $item['quantity'] === $quantity;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:381]: Shouldn't you first find item with a given product name and only then check if it has a proper quantity? Because with this implementation, this method may not work with more than one item in cart

File: features/cart/shopping_cart/resolving_cart_order_items.feature (hunk 8 above)
Comment (arti0090) [reply features/cart/shopping_cart/resolving_cart_order_items.feature:1]: File and feature name has been changed

**Review by GSadee — APPROVED** [review]

**Review by AdamKasp — APPROVED** [review]

**Review by GSadee — COMMENTED** [review]
You should rebase the PR with 1.8 to fix failing build

**Comment by GSadee** [comment]
Thanks, @arti0090! :tada:
